### PR TITLE
Move retry into TokenExchanger implementation

### PR DESCRIPTION
**Pull request by howardjohn**
Broken out of https://github.com/istio/istio/pull/29688. Current `secretcache` and `TokenExchanger` are overloaded, and deals with details they shouldn't. This refactors the TokenExchanger interface to stop leaking HTTP details (stop returning http codes), and `secretcache` to stop leaking sts plugin details (handling http retries). Instead, the retries are moved into the token exchanger directly.

In another PR I will split out the code to do the same for CSR retries (already done in the big PR), then the logic in secret cache is much simpler since there is no retries at all.

This should have roughly the same coverage as before; previously the secretcache tested retries and now the sts client tests directly test these. Logging and metrics for these retries are preserved.

### CODE REVIEW

File: security/pkg/monitoring/monitoring.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package monitoring
+
+import "istio.io/pkg/monitoring"
+
+var (
+	RequestType = monitoring.MustCreateLabel("request_type")
```
Comment (JimmyCYJ) [inline security/pkg/monitoring/monitoring.go:20]: Could you please add a short comment for RequestType?

File: security/pkg/nodeagent/cache/mock/secretcache_mock.go [hunk 2]
```diff
@@ -103,31 +102,15 @@ func (c *CAClient) CSRSign(ctx context.Context, csrPEM []byte, token string, cer
...
-		errors:          errors,
-	}
+// NewMockTokenExchangeServer creates an instance of TokenExchangeServer.
+func NewMockTokenExchangeServer() *TokenExchangeServer {
+	return &TokenExchangeServer{}
 }
 
 // ExchangeToken returns a dumb token or errors depending on the settings.
-func (s *TokenExchangeServer) ExchangeToken(context.Context, security.CredFetcher, string, string) (string, time.Time, int, error) {
-	s.errorCountMutex.Lock()
```
Comment (JimmyCYJ) [inline security/pkg/nodeagent/cache/mock/secretcache_mock.go:124]: Could you keep this logic? I think this is to simulate failures during token exchange, and verify that the retry logic works.

File: security/pkg/nodeagent/cache/secretcache.go [hunk 3]
```diff
@@ -850,14 +829,14 @@ func (sc *SecretCache) getExchangedToken(ctx context.Context, k8sJwtToken string
 		cacheLog.Errorf("Found more than one plugin for %s", logPrefix)
 		return "", fmt.Errorf("found more than one plugin")
 	}
-	exchangedTokens, err := sc.sendRetriableRequest(ctx, nil, k8sJwtToken,
-		ConnKey{ConnectionID: "", ResourceName: ""}, false)
-	if err != nil || len(exchangedTokens) == 0 {
+	p := sc.configOptions.TokenExchangers[0]
```
Comment (JimmyCYJ) [inline security/pkg/nodeagent/cache/secretcache.go:832]: Shall we check if len(sc.configOptions.TokenExchangers) == 0, and return k8sJwtToken with an error message?

File: security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go [hunk 4]
```diff
@@ -57,67 +53,86 @@ type federatedTokenResponse struct {
...
+func retryable(code int) bool {
+	return code >= 500 && !(code == 501 || code == 505 || code == 511)
+}
+
+func (p *Plugin) requestWithRetry(reqBytes []byte) ([]byte, error) {
+	attempts := 0
+	var lastError error
+	for attempts < 5 {
+		attempts++
+		req, err := http.NewRequest("POST", SecureTokenEndpoint, bytes.NewBuffer(reqBytes))
```
Comment (JimmyCYJ) [inline security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go:82]: Shall we move the req generation above the for loop?

File: security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient_test.go [hunk 5]
```diff
@@ -15,33 +15,79 @@
...
+				return fmt.Errorf("expected retries, got %v", g)
+			}
+			return nil
+		}, retry.Timeout(time.Second*5))
+	})
+
+}
+
+func TestMain(m *testing.M) {
+	leak.CheckMain(m)
```
Comment (JimmyCYJ) [inline security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient_test.go:92]: Nice to learn that we can check goroutines leak.

File: security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go (hunk 4 above)
Comment (howardjohn) [reply security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go:82]: This is actually required, an http request cannot be re-used

**Comment by JimmyCYJ** [comment]
Thanks for the refactoring. LGTM overall. cc @myidpt who introduced the original retry logic.

File: security/pkg/nodeagent/cache/mock/secretcache_mock.go (hunk 2 above)
Comment (howardjohn) [reply security/pkg/nodeagent/cache/mock/secretcache_mock.go:124]: This doesn't make sense anymore. Before, the test was: "secretCache calls flaky TokenExchanger, verify retries work". But now Secret cache does *not* retry. Instead, the TokenExchanger itself retries. This is why I have removed this part of the test, but added a test in the TokenExchanger which tests the retries.

File: security/pkg/monitoring/monitoring.go (hunk 1 above)
Comment (howardjohn) [reply security/pkg/monitoring/monitoring.go:20]: Sure, this is copied from the old secretcache package by the way, so its now a new metric. Just moved to a common location.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 3 above)
Comment (howardjohn) [reply security/pkg/nodeagent/cache/secretcache.go:832]: We already do a few lines above (keeping existing logic):
```go
	if sc.configOptions.TokenExchangers == nil || len(sc.configOptions.TokenExchangers) == 0 {
		cacheLog.Debugf("Return k8s token for %s", logPrefix)
		return k8sJwtToken, nil
	}
```

File: security/pkg/monitoring/monitoring.go (hunk 1 above)
Comment (howardjohn) [reply security/pkg/monitoring/monitoring.go:20]: Added comment, thanks

File: security/pkg/nodeagent/cache/mock/secretcache_mock.go (hunk 2 above)
Comment (JimmyCYJ) [reply security/pkg/nodeagent/cache/mock/secretcache_mock.go:124]: I see, thanks!

File: security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go (hunk 4 above)
Comment (JimmyCYJ) [reply security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go:82]: Cool. Thanks.

**Review by JimmyCYJ — APPROVED** [review]

**Comment by howardjohn** [comment]
/retest

unrelated flake, I am already working on fixing

File: security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go [hunk 6]
```diff
@@ -57,67 +53,86 @@ type federatedTokenResponse struct {
...
-	if err != nil {
-		stsClientLog.Errorf("Failed to get SystemCertPool: %v", err)
-		return nil
-	}
-	return Plugin{
-		hTTPClient: &http.Client{
-			Timeout: httpTimeOutInSec * time.Second,
-			Transport: &http.Transport{
-				TLSClientConfig: &tls.Config{
-					RootCAs: caCertPool,
```
Comment (JimmyCYJ) [inline security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go:75]: Forgot to ask, is it still using TLS or plain text now?

Comment (howardjohn) [reply security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient.go:75]: Ah forgot I changed this; yes, the behavior is the same, the http default transport will use TLS whenever `https://` scheme is present and default to system root ca pool (like we manually set up here).

So should be identical behavior

**Pull request merged**